### PR TITLE
Roles in user

### PR DESCRIPTION
**Pull request by lhaskins**
Adds Roles to the User Resource output displayed.

connects to https://github.com/operable/cog/issues/514

### CODE REVIEW

File: lib/cogctl/actions/users.ex [hunk 1]
```diff
@@ -9,21 +9,21 @@ defmodule Cogctl.Actions.Users do
...
     do: with_authentication(endpoint, &do_list/1)
 
-  defp generate_table_row(username, nil, nil), do: [username, ""]
-  defp generate_table_row(username, first, nil), do: [username, first]
-  defp generate_table_row(username, nil, last), do: [username, last]
-  defp generate_table_row(username, first, last), do: [username, first <> " " <> last]
+  defp generate_table_row(username, nil, nil, email), do: [username, "", email]
+  defp generate_table_row(username, first, nil, email), do: [username, first, email]
+  defp generate_table_row(username, nil, last, email), do: [username, last, email]
+  defp generate_table_row(username, first, last, email), do: [username, first <> " " <> last, email]
```
Comment (christophermaier) [inline lib/cogctl/actions/users.ex]: Might be good to break the name creation out into a separate function that just takes `first` and `last`. That's actually what most of this `generate_table_row` function is doing; breaking it apart makes it easier to see what's going on.

File: lib/cogctl/actions/users.ex [hunk 2]
```diff
@@ -9,21 +9,21 @@ defmodule Cogctl.Actions.Users do
...
 
   defp do_list(endpoint) do
-    case CogApi.HTTP.Users.index(endpoint) do
+    case CogApi.HTTP.Client.user_index(endpoint) do
       {:ok, users} ->
         user_attrs = for user <- users do
-          generate_table_row(user.username, user.first_name, user.last_name)
+          generate_table_row(user.username, user.first_name, user.last_name, user.email_address)
```
Comment (christophermaier) [inline lib/cogctl/actions/users.ex]: Actually, why doesn't `generate_table_row` just take a `user`?

File: lib/cogctl/actions/users/info.ex [hunk 3]
```diff
@@ -1,41 +1,50 @@
...
   use Cogctl.Action, "users info"
-  alias Cogctl.Table
+  import Cogctl.Actions.Users.Util, only: [render: 4]
 
   def option_spec do
     [{:user, :undefined, :undefined, {:string, :undefined}, 'User username (required)'}]
   end
 
+  def run(options, _args, _config, %{token: nil}=endpoint),
+    do: with_authentication(endpoint, &run(options, nil, nil, &1))
```
Comment (christophermaier) [inline lib/cogctl/actions/users/info.ex]: Is this ever legitimately called with `endpoint` arguments that have a token and with one that doesn't? If not, this construction is a little awkward to read. For instance, it took me a second to realize that "No, we didn't just accidentally stop authenticating this API endpoint call".

File: lib/cogctl/actions/users/info.ex [hunk 4]
```diff
@@ -1,41 +1,50 @@
...
   def run(options, _args, _config, endpoint) do
-    with_authentication(endpoint,
-                        &do_info(&1, :proplists.get_value(:user, options)))
-  end
-
-  defp do_info(_endpoint, :undefined) do
-    display_arguments_error
+    case :proplists.get_value(:user, options) do
+      :undefined ->
+        display_arguments_error
```
Comment (christophermaier) [inline lib/cogctl/actions/users/info.ex]: I think you'll want to update this to take advantage of changes @mpeck made re: error messages.

File: lib/cogctl/actions/users/info.ex [hunk 5]
```diff
@@ -1,41 +1,50 @@
...
 
-        user_attr = for {title, attr} <- [{"ID", "id"}, {"Username", "username"}, {"First Name", "first_name"}, {"Last Name", "last_name"}, {"Email", "email_address"}] do
-          [title, user[attr]]
+        user_attr = for {title, attr} <- [{"ID", :id}, {"Username", :username}, {"First Name", :first_name}, {"Last Name", :last_name}, {"Email", :email_address}] do
+          [title, Map.fetch!(user, attr)]
         end
 
-        groups = Enum.map(user["groups"], fn(membership) ->
-                   [membership["name"], membership["id"]]
+        groups = Enum.map(user.groups, fn(membership) ->
```
Comment (christophermaier) [inline lib/cogctl/actions/users/info.ex]: `group` seems like a more natural name than `membership` here.

File: lib/cogctl/actions/users/info.ex [hunk 6]
```diff
@@ -1,41 +1,50 @@
...
                  end)
 
-        display_output("""
-                       #{Table.format(user_attr, false)}
+        role_attr = Enum.reduce(user.groups, [], fn(membership, roles) ->
+                      roles ++ Enum.map(membership.roles, fn(role) ->
+                          permissions = role.permissions |> build_permissions |> Enum.join(",")
+                          [role.name, role.id, permissions]
+                      end)
+                end)
```
Comment (christophermaier) [inline lib/cogctl/actions/users/info.ex]: Probably a good idea to break this up, since there's a lot going on.

It doesn't look like you're separating roles out by the groups they're attached to, so this will repeat entries if you're, say, in two groups that each are granted the same role.

Try getting the unique roles first, then doing an `Enum.map` (because this is fundamentally a `map` operation, not a `reduce`) to generate the table rows.

Comment (christophermaier) [inline lib/cogctl/actions/users/info.ex]: Incidentally, we should consistently refer to things that are going to be "table rows". I've seen `*_attr`, `*_row`, and `groups`. They're all fundamentally the same thing, so we should name them accordingly.

File: lib/cogctl/actions/users/info.ex [hunk 7]
```diff
@@ -1,41 +1,50 @@
...
-                       """ |> String.rstrip)
+
+        render(user_attr, [["NAME", "ID"]] ++ groups, [["NAME", "ID", "PERMISSIONS"]] ++ role_attr, true)
       {:error, error} ->
-        display_error(error["errors"])
+        display_error(error)
     end
   end
+
+  defp build_permissions(permissions) do
```
Comment (christophermaier) [inline lib/cogctl/actions/users/info.ex]: I'd favor a function called `permission_name` that just generates the string and the use 

```elixir
|> Enum.map(&permission_name/1)
```
at the call site.

The name `build_permissions` is vague, and obscures the fact that we're just formatting a name (also, what are you building if you're already passing the function a list of permissions?). Hiding the `map` call inside a more complicated function doesn't gain anything in this case.

File: lib/cogctl/actions/users/info.ex [hunk 8]
```diff
@@ -1,41 +1,50 @@
...
+                          permissions = role.permissions |> build_permissions |> Enum.join(",")
+                          [role.name, role.id, permissions]
+                      end)
+                end)
 
-                       Groups
-                       #{Table.format([["NAME", "ID"]] ++ groups, true)}
-                       """ |> String.rstrip)
+
+        render(user_attr, [["NAME", "ID"]] ++ groups, [["NAME", "ID", "PERMISSIONS"]] ++ role_attr, true)
```
Comment (christophermaier) [inline lib/cogctl/actions/users/info.ex]: It'd also be a good idea to put the group and role table creation into their own functions and encapsulate the header row and data extraction in one place.

File: lib/cogctl/actions/users/info.ex [hunk 9]
```diff
@@ -6,38 +6,30 @@ defmodule Cogctl.Actions.Users.Info do
...
+      {:ok, params} ->
+        with_authentication(endpoint, &do_info(&1, params))
+      {:error, {:missing_params, missing_args}} ->
+        display_arguments_error(missing_args)
     end
   end
 
-  defp do_info(endpoint, username) do
-    case CogApi.HTTP.Client.user_show(endpoint, %{username: username}) do
+  defp do_info(endpoint, params) do
```
Comment (christophermaier) [inline lib/cogctl/actions/users/info.ex]: Why bother passing in `params` when it's really the user that you care about?

File: lib/cogctl/actions/users/info.ex [hunk 10]
```diff
@@ -47,4 +39,18 @@ defmodule Cogctl.Actions.Users.Info do
     Enum.map(permissions, fn(perm) ->
       "#{perm["namespace"]}:#{perm["name"]}" end)
   end
+
+  defp get_roles_from_groups(groups) do
+    Enum.reduce(groups, [], fn(group, roles) ->
```
Comment (christophermaier) [inline lib/cogctl/actions/users/info.ex]: Use `Enum.flat_map` instead of `Enum.reduce`.

File: lib/cogctl/actions/users/util.ex [hunk 11]
```diff
@@ -8,9 +8,16 @@ defmodule Cogctl.Actions.Users.Util do
 
   def render(user_info, group_info, role_info, sort) do
     Table.format(user_info, false)
-        <> "\n\nGroups:\n" <> Table.format(group_info, sort)
-        <> "\n\nRoles:\n" <> Table.format(role_info, sort)
+        <> "\n\nGroups:\n" <> render_groups(group_info, sort)
+        <> "\n\nRoles:\n" <> render_roles(role_info, sort)
```
Comment (vanstee) [inline lib/cogctl/actions/users/util.ex]: It's always a little tough to parse literal `\n` in here. Could be cool to use here docs like:

```
display_output("""
  #{Table.format(user_info, false)}

  Groups:
  #{render_groups(group_info, sort)}

  Roles:
  #{render_roles(role_info, sort)}
""")
```

File: lib/cogctl/actions/users/info.ex [hunk 12]
```diff
@@ -47,4 +39,18 @@ defmodule Cogctl.Actions.Users.Info do
...
+                roles ++ get_roles(group.roles)
+    end)
+    |> Enum.uniq
+  end
+
+  defp get_roles(roles) do
```
Comment (christophermaier) [inline lib/cogctl/actions/users/info.ex]: `get_roles_from_groups` and this `get_roles` function are still entangling too many operations. You need to first find the unique roles, then process them as table rows.

File: lib/cogctl/actions/users/util.ex (hunk 11 above)
Comment (vanstee) [inline lib/cogctl/actions/users/util.ex]: Also not sure if there's an easy fix for this, but @imbriaco brought up a good point yesterday about out data from multiple lists. I wonder if we should only render groups and/or roles when a specific flag is present like `cogctl users info vanstee --groups`. Another option would be to move that stuff to something like `cogctl groups --user=vanstee`.

Not sure if that's a change we should make in this PR or not though.

Comment (mpeck) [inline lib/cogctl/actions/users/util.ex]: I agree with @vanstee. I like an optional flag like `--details` or `--groups`. But I think that might be for another PR.

File: lib/cogctl/actions/users.ex [hunk 13]
```diff
@@ -9,21 +9,25 @@ defmodule Cogctl.Actions.Users do
...
-  defp generate_table_row(username, nil, last), do: [username, last]
-  defp generate_table_row(username, first, last), do: [username, first <> " " <> last]
+  defp generate_table_row(user) do
+    [user.username, get_name(user.first_name, user.last_name), user.email_address]
+  end
+
+  defp get_name(nil, nil), do: ""
+  defp get_name(nil, last), do: last
+  defp get_name(first, nil), do: first
+  defp get_name(first, last), do: first <> " " <> last
```
Comment (mpeck) [inline lib/cogctl/actions/users.ex]: As an alternate you could do,
```
Enum.reject([first_name, last_name], &is_nil/1)
|> Enum.join(" ")
```

File: lib/cogctl/actions/users/info.ex (hunk 12 above)
Comment (mpeck) [inline lib/cogctl/actions/users/info.ex]: Might be a good use for `get_in`?
```
all = fn(:get, data, next) ->
  Enum.flat_map(data, &next.(Map.delete(&1, :__struct__)))
end

roles = get_in(user.groups, [all, :roles])
|> Enum.uniq
```
Then in your table rendering logic, something like:
```
Enum.map(roles, fn(role) ->
  permissions = Enum.map_join(role.permissions, ",", &"#{&1["namespace"]}:#{&1["name"]}")
  [role.name, role.id, permissions]
end)
```

File: test/cogctl_test.exs [hunk 14]
```diff
@@ -122,10 +122,24 @@ defmodule CogctlTest do
...
-    Groups
-    NAME       ID
-    cog-admin  .*
+    assert run("cogctl users info admin --groups") =~ ~r"""
+    ID          .*
+    Username    admin
+    First Name  Cog
+    Last Name   Administrator
+    Email       cog@localhost
+    Groups:     cog-admin
```
Comment (christophermaier) [inline test/cogctl_test.exs]: Does it work properly if you're in multiple groups, or have multiple roles?

Comment (lhaskins) [inline test/cogctl_test.exs]: Yes, I'll add another test so that's clear.

**Comment by mpeck** [comment]
:+1:
